File: stdlib/js/hull/uuid.js
/**
 * @file hull:uuid
 * @module hull:uuid
 * @description RFC 9562 UUID generation (v4 random, v7 time-ordered).
 *   Lua parity: `hull.uuid`.
 *
 * Both variants are 36-char canonical strings
 * (`xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx`, `M` the version nibble, `N` the
 * variant). Built on `crypto.random` (CSPRNG) + `time.nowMs`; no new authority.
 * Prefer these over ad-hoc ids from `crypto.base64urlEncode(crypto.random(16))`:
 * canonical, externally interoperable, and v7 is lexically sortable.
 *
 * @license AGPL-3.0-or-later
 * @example
 *   import { uuid } from "hull:uuid";
 *   const id = uuid.v7();   // time-ordered; ideal for DB primary keys
 *   const r  = uuid.v4();   // fully random
 */

import { crypto } from "hull:crypto";
import { time } from "hull:time";

// Byte -> 2-hex-char lookup, built once.
const HEX = new Array(256);
for (let i = 0; i < 256; i++) HEX[i] = (i + 0x100).toString(16).slice(1);

function fmt(b) {
    return HEX[b[0]] + HEX[b[1]] + HEX[b[2]] + HEX[b[3]] + "-" +
           HEX[b[4]] + HEX[b[5]] + "-" + HEX[b[6]] + HEX[b[7]] + "-" +
           HEX[b[8]] + HEX[b[9]] + "-" +
           HEX[b[10]] + HEX[b[11]] + HEX[b[12]] + HEX[b[13]] + HEX[b[14]] + HEX[b[15]];
}

/**
 * Random (version 4) UUID.
 * @returns {string} 36-char canonical UUID.
 */
function v4() {
    const b = new Uint8Array(crypto.random(16));
    b[6] = (b[6] & 0x0f) | 0x40;   // version 4
    b[8] = (b[8] & 0x3f) | 0x80;   // variant 10
    return fmt(b);
}

/**
 * Time-ordered (version 7) UUID: 48-bit big-endian Unix-ms timestamp followed
 * by 74 random bits. Lexically sortable by creation time - good for database
 * primary keys.
 * @returns {string} 36-char canonical UUID.
 */
function v7() {
    const b = new Uint8Array(16);
    // 48-bit ms timestamp, big-endian. ms exceeds 2^32, so peel bytes by
    // division (JS bitwise ops are 32-bit and would overflow).
    let t = time.nowMs();
    for (let i = 5; i >= 0; i--) { b[i] = t % 256; t = Math.floor(t / 256); }
    const r = new Uint8Array(crypto.random(10));
    for (let i = 0; i < 10; i++) b[6 + i] = r[i];
    b[6] = (b[6] & 0x0f) | 0x70;   // version 7
    b[8] = (b[8] & 0x3f) | 0x80;   // variant 10
    return fmt(b);
}

export const uuid = { v4, v7 };
export default uuid;

File: stdlib/lua/hull/uuid.lua
--- RFC 9562 UUID generation (v4 random, v7 time-ordered).
--
-- Both variants are 36-char canonical strings
-- (`xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx`, `M` the version nibble, `N` the
-- variant). Built on `crypto.random` (CSPRNG) + `time.now_ms`; no new
-- authority. Apps that previously minted ad-hoc ids from
-- `crypto.base64url_encode(crypto.random(16))` should prefer these - they are
-- canonical, interoperable with external systems, and v7 is lexically sortable.
--
-- @module hull.uuid
-- @license AGPL-3.0-or-later
-- @usage
--   local uuid = require("hull.uuid")
--   local id = uuid.v7()   -- time-ordered; ideal for DB primary keys
--   local r  = uuid.v4()   -- fully random

local crypto = require("hull.crypto")
local time = require("hull.time")

local uuid = {}

local FMT = "%02x%02x%02x%02x-%02x%02x-%02x%02x-%02x%02x-%02x%02x%02x%02x%02x%02x"

--- Random (version 4) UUID.
-- @treturn string  36-char canonical UUID.
function uuid.v4()
    local b = { crypto.random(16):byte(1, 16) }
    b[7] = (b[7] & 0x0f) | 0x40   -- version 4
    b[9] = (b[9] & 0x3f) | 0x80   -- variant 10
    return string.format(FMT, table.unpack(b))
end

--- Time-ordered (version 7) UUID: 48-bit big-endian Unix-ms timestamp followed
-- by 74 random bits. Lexically sortable by creation time - good for database
-- primary keys (better index locality than v4).
-- @treturn string  36-char canonical UUID.
function uuid.v7()
    local ms = time.now_ms()
    local r = { crypto.random(10):byte(1, 10) }
    local b = {
        (ms >> 40) & 0xff, (ms >> 32) & 0xff, (ms >> 24) & 0xff,
        (ms >> 16) & 0xff, (ms >> 8) & 0xff, ms & 0xff,
        r[1], r[2], r[3], r[4], r[5], r[6], r[7], r[8], r[9], r[10],
    }
    b[7] = (b[7] & 0x0f) | 0x70   -- version 7
    b[9] = (b[9] & 0x3f) | 0x80   -- variant 10
    return string.format(FMT, table.unpack(b))
end

return uuid
